### PR TITLE
Queue refactor

### PR DESCRIPTION
**Pull request by MindFlavor**
Building on https://github.com/Azure/azure-sdk-for-rust/pull/109, this is a draft of the same changes on azure queue storage.

This PR follows the same patterns and proposes some new types (`Timeout`, `ClienrtRequestId` and `Metadata`), removes the trait functions (now simply struct functions) for the builder(s).

I am not happy with the `Metadata` struct so suggestions are welcome: for the time being it's a string owning hashmap but maybe a &str based one would be better? A Cow to give the user liberty to choose?

### CODE REVIEW

File: sdk/core/src/client_request_id.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    }
+}
+
+impl<'a> AddAsHeader for ClientRequestId<'a> {
+    fn add_as_header(&self, builder: Builder) -> Builder {
+        builder.header(CLIENT_REQUEST_ID, self.0.as_ref())
+    }
+}
+
+impl<'a> AddAsHeader for Option<ClientRequestId<'a>> {
```
Comment (rylev) [inline sdk/core/src/client_request_id.rs:21]: Is this really needed when we'll be using the `add_header` helper function?

Comment (rylev) [inline sdk/core/src/client_request_id.rs:21]: Is this really needed when we'll be using the `add_header` helper function?

File: sdk/core/src/metadata.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+#[derive(Debug, Clone)]
+pub struct Metadata(HashMap<String, String>);
+
+impl Metadata {
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+
+    pub fn as_mut(&mut self) -> &mut HashMap<String, String> {
```
Comment (rylev) [inline sdk/core/src/metadata.rs:13]: Maybe we can provide an `insert` for convenience since I imagine 99% of the time people will call `as_mut().insert(key, value)`

File: sdk/storage/src/queue/requests/create_queue_builder.rs [hunk 3]
```diff
@@ -123,16 +86,18 @@ where
...
 
         debug!("uri == {}", url);
 
         let perform_request_response = self.queue_client.storage_client().perform_request(
             url.as_str(),
             &http::Method::PUT,
             &|mut request| {
-                request = ClientRequestIdOption::add_header(&self, request);
-                request = MetadataOption::add_header(&self, request);
+                request = AddAsHeader::add_as_header(self.client_request_id(), request);
```
Comment (rylev) [inline sdk/storage/src/queue/requests/create_queue_builder.rs:97]: Can we use the `add_header` helper function here?

**Review by rylev — COMMENTED** [review]
Looking good so far. I have some small comments.

File: sdk/core/src/client_request_id.rs (hunk 1 above)
Comment (MindFlavor) [reply sdk/core/src/client_request_id.rs:21]: yes it's probably best to the helper function 👍 !

File: sdk/storage/src/queue/requests/set_queue_metadata_builder.rs [hunk 4]
```diff
@@ -86,7 +87,7 @@ where
             url.as_str(),
             &http::Method::PUT,
             &|mut request| {
-                request = AddAsHeader::add_as_header(self.client_request_id(), request);
+                request = add_header(self.client_request_id(), request);
                 request = AddAsHeader::add_as_header(&self.metadata(), request);
```
Comment (rylev) [inline sdk/storage/src/queue/requests/set_queue_metadata_builder.rs:91]: Why not this one as well?

Comment (MindFlavor) [reply sdk/storage/src/queue/requests/set_queue_metadata_builder.rs:91]: `metadata` is not `Option` so we cannot use the helper function. Should we create a non option variant? I would be a one liner so I thought better to call the trait function directly.

Comment (rylev) [reply sdk/storage/src/queue/requests/set_queue_metadata_builder.rs:91]: It's up to you. In cosmos I wrap the ones that are not options in `Some`. The compiler most likely can see through this and will skip the if check `add_header`.

Comment (MindFlavor) [reply sdk/storage/src/queue/requests/set_queue_metadata_builder.rs]: I think than using wrapping into a fake `Option` for the sake of saving a function makes for confusing code. 

I suggest we have two functions instead: `add_optional_header` and `add_mandatory_header`. Their names are self-explanatory even to a casual reader.

File: sdk/cosmos/src/requests/create_document_builder.rs [hunk 5]
```diff
@@ -201,15 +201,15 @@ impl<'a, 'b> CreateDocumentBuilder<'a, 'b, Yes> {
...
-        req = crate::headers::add_header(self.if_match_condition(), req);
-        req = crate::headers::add_header(self.if_modified_since(), req);
-        req = crate::headers::add_header(self.user_agent(), req);
-        req = crate::headers::add_header(self.activity_id(), req);
-        req = crate::headers::add_header(self.consistency_level(), req);
-        req = crate::headers::add_header(Some(self.partition_keys()), req);
-        req = crate::headers::add_header(Some(self.is_upsert()), req);
-        req = crate::headers::add_header(Some(self.indexing_directive()), req);
-        req = crate::headers::add_header(Some(self.allow_tentative_writes()), req);
+        req = azure_core::headers::add_optional_header(&self.if_match_condition(), req);
```
Comment (MindFlavor) [inline sdk/cosmos/src/requests/create_document_builder.rs:204]: I think this is better.

File: sdk/core/src/client_request_id.rs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+use crate::headers::*;
+use crate::AddAsHeader;
+use http::request::Builder;
+use std::borrow::Cow;
+
+#[derive(Debug, Clone)]
+pub struct ClientRequestId<'a>(Cow<'a, str>);
```
Comment (rylev) [inline sdk/core/src/client_request_id.rs:7]: I personally don't see a point of using `Cow` here. If this is simply to save users the convenience of not writting `&` before an owned string that they want to pass to `ClientRequestId::new`, that is very much not worth the complexity of the type signature that they will see when reading the docs.

File: sdk/core/src/headers/mod.rs [hunk 7]
```diff
@@ -1,7 +1,24 @@
...
+    mut builder: Builder,
+) -> Builder {
+    if let Some(item) = item {
+        builder = item.add_as_header(builder);
+    }
+    builder
+}
+
+#[must_use]
+pub fn add_mandatory_header<T: crate::AddAsHeader>(item: &T, builder: Builder) -> Builder {
```
Comment (rylev) [inline sdk/core/src/headers/mod.rs:18]: We should switch Cosmos to use these methods, but of course we can do that in another PR.

File: sdk/core/src/metadata.rs [hunk 8]
```diff
@@ -0,0 +1,47 @@
+use crate::AddAsHeader;
+use http::request::Builder;
+use std::borrow::Cow;
+use std::collections::HashMap;
+
+#[derive(Debug, Clone)]
+pub struct Metadata<'a>(HashMap<Cow<'a, str>, Cow<'a, str>>);
```
Comment (rylev) [inline sdk/core/src/metadata.rs:7]: Same point here about `Cow`. This makes the type harder to understand and thus harder to use.

File: sdk/core/src/metadata.rs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        s
+    }
+}
+
+impl<'a> AddAsHeader for &Metadata<'a> {
+    fn add_as_header(&self, builder: Builder) -> Builder {
+        let mut builder = builder;
+
+        for (key, val) in self.0.iter() {
+            builder = builder.header(&format!("x-ms-meta-{}", key) as &str, val as &str);
```
Comment (rylev) [inline sdk/core/src/metadata.rs:42]: I don't believe the first `as &str` is necessary.

File: sdk/core/src/modify_conditions.rs [hunk 10]
```diff
@@ -13,7 +13,7 @@ pub enum IfSinceCondition {
 }
 
 impl IfSinceCondition {
-    pub(crate) fn add_header(&self, builder: Builder) -> Builder {
```
Comment (rylev) [inline sdk/core/src/modify_conditions.rs:16]: Why rename these? We can probably just get rid of them entirely since we can use their implementation of `AddAsHeader` instead, no?

File: sdk/storage/examples/queue_create.rs [hunk 11]
```diff
@@ -25,20 +24,36 @@ async fn main() -> Result<(), Box<dyn Error>> {
 
     // this step is optional but here we show
     // how to add metadata to a new queue.
-    let mut hm = HashMap::new();
-    hm.insert("source", "azure-sdk-for-rust");
+    let mut metadata = Metadata::new();
+    metadata
+        .as_mut()
```
Comment (rylev) [inline sdk/storage/examples/queue_create.rs:29]: Use `insert` directly instead of `as_mut`

File: sdk/storage/examples/queue_create.rs [hunk 12]
```diff
@@ -25,20 +24,36 @@ async fn main() -> Result<(), Box<dyn Error>> {
...
+        .insert("source".into(), "azure-sdk-for-rust".into());
```
Comment (rylev) [inline sdk/storage/examples/queue_create.rs:30]: This is a good example of where `Cow` makes things more complicated

File: sdk/storage/src/queue/requests/get_messages_builder.rs [hunk 13]
```diff
@@ -107,55 +90,43 @@ where
...
-impl<'a, C> TimeoutSupport for GetMessagesBuilder<'a, C>
+// methods callable only when every mandatory field has been filled
+impl<'a, C> GetMessagesBuilder<'a, C>
 where
     C: Client + Clone,
 {
-    type O = Self;
+    pub fn queue_client(&self) -> &'a QueueClient<C> {
+        self.queue_client
+    }
```
Comment (rylev) [inline sdk/storage/src/queue/requests/get_messages_builder.rs:100]: I hope like in #111 we can get rid of these getter methods. They're not really needed.

File: sdk/storage/src/queue/requests/set_queue_metadata_builder.rs [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        self.queue_client
+    }
+
+    pub fn metadata(&self) -> &'a Metadata {
+        self.metadata
+    }
+
+    pub fn timeout(&self) -> &Option<Timeout> {
+        &self.timeout
+    }
```
Comment (rylev) [inline sdk/storage/src/queue/requests/set_queue_metadata_builder.rs:50]: Same here. Can we just remove these?

File: sdk/core/src/prelude.rs [hunk 15]
```diff
@@ -2,32 +2,33 @@ pub use crate::ba512_range::BA512Range;
...
+    NextMarkerOption, NextMarkerSupport, PageBlobLengthRequired, PageBlobLengthSupport,
+    PrefixOption, PrefixSupport, ProposedLeaseIdOption, ProposedLeaseIdRequired,
+    ProposedLeaseIdSupport, RangeOption, RangeRequired, RangeSupport,
+    SequenceNumberConditionOption, SequenceNumberConditionSupport, SequenceNumberOption,
+    SequenceNumberSupport, SnapshotOption, SnapshotRequired, SnapshotSupport,
     SourceContentMD5Option, SourceContentMD5Support, SourceLeaseIdOption, SourceLeaseIdSupport,
-    SourceUrlRequired, SourceUrlSupport, StoredAccessPolicy, StoredAccessPolicyList, TimeoutOption,
-    TimeoutSupport, UserAgent, UserAgentSupport, EMPTY_BODY,
+    SourceUrlRequired, SourceUrlSupport, StoredAccessPolicy, StoredAccessPolicyList, Timeout,
+    TimeoutOption, TimeoutSupport, UserAgent, UserAgentSupport, EMPTY_BODY,
```
Comment (rylev) [inline sdk/core/src/prelude.rs:33]: Can we delete `TimeoutOption` and any other traits that we're no longer using? Eventually most of this file can be deleted.

**Review by rylev — CHANGES_REQUESTED** [review]
I really love the direction of this PR, but I'd like to see some of the old cruft that's no longer used deleted, and I still see some things that are more complicated than they need to be.

File: sdk/core/src/client_request_id.rs (hunk 6 above)
Comment (MindFlavor) [reply sdk/core/src/client_request_id.rs:7]: Good point! Changed!

File: sdk/core/src/headers/mod.rs (hunk 7 above)
Comment (MindFlavor) [reply sdk/core/src/headers/mod.rs:18]: It's included in this PR 👍

File: sdk/core/src/metadata.rs (hunk 8 above)
Comment (MindFlavor) [reply sdk/core/src/metadata.rs:7]: I am on the fence on whether to accept `&str` or a `String`. `&str` means the `Metadata` struct cannot escape the owning function easily while forcing `string` means potentially copying hundreds of strings around. 

I'll go with `String` for ergonomics.

File: sdk/core/src/modify_conditions.rs (hunk 10 above)
Comment (MindFlavor) [reply sdk/core/src/modify_conditions.rs:16]: Yes I've just renamed these in bulk with a brutal `sed` script 🪓. I haven't bothered to correct them since they are obsolete and will be removed as soon as the remaining code no longer references them.

File: sdk/storage/src/queue/requests/get_messages_builder.rs (hunk 13 above)
Comment (MindFlavor) [reply sdk/storage/src/queue/requests/get_messages_builder.rs:100]: Yes I wholeheartedly agree 👍 👍 !

File: sdk/storage/src/queue/requests/set_queue_metadata_builder.rs (hunk 14 above)
Comment (MindFlavor) [reply sdk/storage/src/queue/requests/set_queue_metadata_builder.rs:50]: Removed!

File: sdk/core/src/prelude.rs (hunk 15 above)
Comment (MindFlavor) [reply sdk/core/src/prelude.rs:33]: Yes as soon as they have been removed from every request builder we'll removed them. The other PR (#119) will address the rest of storage request builders (I will have to merge this one first because it contains the `Timeout` struct).

File: sdk/core/src/metadata.rs (hunk 8 above)
Comment (rylev) [reply sdk/core/src/metadata.rs]: For everything else we're taking &str. I'm fine with going with `String`. We might want to think about why `Metadata` is special and takes `String`. Perhaps more types should take `String` instead of `&str`. Something to think about....

Comment (MindFlavor) [reply sdk/core/src/metadata.rs]: My reasoning is: for something that does not make sense outside a request (like `client_request_id`) `&str` is fine. This is because you use it on a request only, and a requests are short lived 99% of the time (so borrowing makes sense).

For everything else `String` is better.

**Review by rylev — APPROVED** [review]
